### PR TITLE
[MIPS] Fix failing to legalize load+call with vector of non-p2 intege…

### LINKED ISSUE #102870: [MIPS] Failure to legalize load+call with vector of non-p2 integer type
```
❯ cat reduced.ll
target datalayout = "e-m:m-p:32:32-i8:8:32-i16:16:32-i64:64-n32-S64"
target triple = "mipsel-unknown-linux-unknown"

define void @_start() {
Entry:
  %0 = load <1 x i80>, ptr null, align 16
  call fastcc void null(<1 x i80> %0)
  ret void
}
❯ llc -O0 reduced.ll
LLVM ERROR: Do not know how to expand this operator's operand!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: llc -O0 reduced.ll
1.      Running pass 'Function Pass Manager' on module 'reduced.ll'.
2.      Running pass 'MIPS DAG->DAG Pattern Instruction Selection' on function '@_start'
 #0 0x00007ff7d6e62b42 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xa62b42)
 #1 0x00007ff7d6e5ff9b SignalHandler(int) Signals.cpp:0:0
 #2 0x00007ff7d5c42990 (/lib/x86_64-linux-gnu/libc.so.6+0x42990)
 #3 0x00007ff7d5c99a1b __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
 #4 0x00007ff7d5c99a1b __pthread_kill_internal ./nptl/pthread_kill.c:78:10
 #5 0x00007ff7d5c99a1b pthread_kill ./nptl/pthread_kill.c:89:10
 #6 0x00007ff7d5c428e6 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #7 0x00007ff7d5c268b7 abort ./stdlib/abort.c:81:7
 #8 0x00007ff7d6b84ced llvm::json::operator==(llvm::json::Value const&, llvm::json::Value const&) (.cold) JSON.cpp:0:0
 #9 0x00007ff7d6d671ce (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x9671ce)
#10 0x00007ff7d786db21 llvm::DAGTypeLegalizer::ExpandIntegerOperand(llvm::SDNode*, unsigned int) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x146db21)
#11 0x00007ff7d78908e9 llvm::DAGTypeLegalizer::run() (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x14908e9)
#12 0x00007ff7d7890fdd llvm::SelectionDAG::LegalizeTypes() (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x1490fdd)
#13 0x00007ff7d79ec534 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x15ec534)
#14 0x00007ff7d79ef178 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x15ef178)
#15 0x00007ff7d79f179a llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (.part.0) SelectionDAGISel.cpp:0:0
#16 0x00007ff7da5fd81e llvm::MipsDAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0x41fd81e)
#17 0x00007ff7d739ab08 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (.part.0) MachineFunctionPass.cpp:0:0
#18 0x00007ff7d701abd9 llvm::FPPassManager::runOnFunction(llvm::Function&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xc1abd9)
#19 0x00007ff7d701af24 llvm::FPPassManager::runOnModule(llvm::Module&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xc1af24)
#20 0x00007ff7d701b92e llvm::legacy::PassManagerImpl::run(llvm::Module&) (/opt/llvm/bin/../lib/libLLVM.so.18.1+0xc1b92e)
#21 0x00005ea99532abeb compileModule(char**, llvm::LLVMContext&) llc.cpp:0:0
#22 0x00005ea995320756 main (/opt/llvm/bin/llc+0x11756)
#23 0x00007ff7d5c28150 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:74:3
#24 0x00007ff7d5c28209 call_init ./csu/../csu/libc-start.c:128:20
#25 0x00007ff7d5c28209 __libc_start_main ./csu/../csu/libc-start.c:347:5
#26 0x00005ea995321415 _start (/opt/llvm/bin/llc+0x12415)
❯ llc --version
LLVM (http://llvm.org/):
  LLVM version 18.1.8
  Optimized build.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: alderlake

  Registered Targets:
    aarch64     - AArch64 (little endian)
    aarch64_32  - AArch64 (little endian ILP32)
    aarch64_be  - AArch64 (big endian)
    amdgcn      - AMD GCN GPUs
    arc         - ARC
    arm         - ARM
    arm64       - ARM64 (little endian)
    arm64_32    - ARM64 (little endian ILP32)
    armeb       - ARM (big endian)
    avr         - Atmel AVR Microcontroller
    bpf         - BPF (host endian)
    bpfeb       - BPF (big endian)
    bpfel       - BPF (little endian)
    csky        - C-SKY
    dxil        - DirectX Intermediate Language
    hexagon     - Hexagon
    lanai       - Lanai
    loongarch32 - 32-bit LoongArch
    loongarch64 - 64-bit LoongArch
    m68k        - Motorola 68000 family
    mips        - MIPS (32-bit big endian)
    mips64      - MIPS (64-bit big endian)
    mips64el    - MIPS (64-bit little endian)
    mipsel      - MIPS (32-bit little endian)
    msp430      - MSP430 [experimental]
    nvptx       - NVIDIA PTX 32-bit
    nvptx64     - NVIDIA PTX 64-bit
    ppc32       - PowerPC 32
    ppc32le     - PowerPC 32 LE
    ppc64       - PowerPC 64
    ppc64le     - PowerPC 64 LE
    r600        - AMD GPUs HD2XXX-HD6XXX
    riscv32     - 32-bit RISC-V
    riscv64     - 64-bit RISC-V
    sparc       - Sparc
    sparcel     - Sparc LE
    sparcv9     - Sparc V9
    spirv       - SPIR-V Logical
    spirv32     - SPIR-V 32-bit
    spirv64     - SPIR-V 64-bit
    systemz     - SystemZ
    thumb       - Thumb
    thumbeb     - Thumb (big endian)
    ve          - VE
    wasm32      - WebAssembly 32-bit
    wasm64      - WebAssembly 64-bit
    x86         - 32-bit X86: Pentium-Pro and above
    x86-64      - 64-bit X86: EM64T and AMD64
    xcore       - XCore
    xtensa      - Xtensa 32
```

### PR DESCRIPTION
**Pull request by yingopq**
…r type

Add a condition to check whether the vector element type is a power of 2.

Fix #102870.

### CODE REVIEW

File: llvm/lib/Target/Mips/MipsISelLowering.cpp [hunk 1]
```diff
@@ -124,7 +124,8 @@ unsigned MipsTargetLowering::getNumRegistersForCallingConv(LLVMContext &Context,
 unsigned MipsTargetLowering::getVectorTypeBreakdownForCallingConv(
     LLVMContext &Context, CallingConv::ID CC, EVT VT, EVT &IntermediateVT,
     unsigned &NumIntermediates, MVT &RegisterVT) const {
-  if (VT.isPow2VectorType()) {
+
```
Comment (nikic) [inline llvm/lib/Target/Mips/MipsISelLowering.cpp:127]: Stray empty line.

File: llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll [hunk 2]
```diff
@@ -0,0 +1,30 @@
+; RUN: llc -mtriple=mips-linux-gnu -mcpu=mips32 < %s | FileCheck %s --check-prefix=MIPS32
+
+define void @call_v1i80() {
```
Comment (nikic) [inline llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll:3]: Add nounwind to suppress cfi.

File: llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+; MIPS32-NEXT:    sll $1, $1, 16
+; MIPS32-NEXT:    or $6, $2, $1
+; MIPS32-NEXT:    addiu	$25, $zero, 0
+; MIPS32-NEXT:    jalr $25
+; MIPS32-NEXT:    srl $4, $3, 16
+; MIPS32-NEXT:    lw $ra, 4($sp) # 4-byte Folded Reload
+; MIPS32-NEXT:    jr $ra
+; MIPS32-NEXT:    addiu $sp, $sp, 8
+Entry:
+  %0 = load <1 x i80>, ptr null, align 16
```
Comment (nikic) [inline llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll:27]: This test case is over-reduced. Does the problem still reproduce if you replace the nulls here with a real pointer and a real function declaration?

File: llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll
Comment (nikic) [inline llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll:1]: I'd add this to the illegal-vectors.ll test, which has a lot of related coverage.

File: llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll (hunk 3 above)
Comment (yingopq) [reply llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll:27]: Yes, this situation you mentioned also can be solved and the reported bug info was same as null pointer situation. I tested the following test case.

```
$ cat 1.ll
target datalayout = "e-m:m-p:32:32-i8:8:32-i16:16:32-i64:64-n32-S64"
target triple = "mipsel-unknown-linux-unknown"

define void @arg_v1i80(<1 x i80> %vec, ptr %p) {
  store <1 x i80> %vec, ptr %p
  ret void
}

define void @_start(ptr %p) nounwind{
Entry:
  %0 = load <1 x i80>, ptr %p, align 16
  call void @arg_v1i80(<1 x i80> %0)
  ret void
}

```

File: llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll
Comment (yingopq) [reply llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll:1]: Need I added `call_v1i80` to this file?

File: llvm/lib/Target/Mips/MipsISelLowering.cpp (hunk 1 above)
Comment (yingopq) [reply llvm/lib/Target/Mips/MipsISelLowering.cpp]: Applied.

File: llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll
Comment (yingopq) [reply llvm/test/CodeGen/Mips/cconv/not-power-of-2-vector.ll:1]: Move test to illegal-vectors.ll done.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**